pkg/scheduler/util/assumecache/assume_cache.go

```diff
@@ -0,0 +1,420 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package assumecache
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"sync"
+
+	"k8s.io/klog/v2"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/client-go/tools/cache"
+)
+
+// AssumeCache is a cache on top of the informer that allows for updating
+// objects outside of informer events and also restoring the informer
+// cache's version of the object. Objects are assumed to be
+// Kubernetes API objects that are supported by [meta.Accessor].
+//
+// Objects can referenced via their key, with [cache.MetaNamespaceKeyFunc]
+// as key function.
+type AssumeCache interface {
+	// Assume updates the object in-memory only.
+	//
+	// The version of the object must be greater or equal to
+	// the current object, otherwise an error is returned.
+	//
+	// Storing an object with the same version is supported
+	// by the assume cache, but suffers from a race: if an
+	// update is received via the informer while such an
+	// object is assumed, it gets dropped in favor of the
+	// newer object from the apiserver.
+	//
+	// Only assuming objects that were returned by an apiserver
+	// operation (Update, Patch) is safe.
+	Assume(obj interface{}) error
+
+	// Restore the informer cache's version of the object.
+	Restore(key string)
+
+	// Get the object by its key.
+	Get(key string) (interface{}, error)
+
+	// GetAPIObj gets the informer cache's version by its key.
+	GetAPIObj(key string) (interface{}, error)
+
+	// List all the objects in the cache.
+	List(indexObj interface{}) []interface{}
+
+	// getImplementation is used internally by [AddTestObject], [UpdateTestObject], [DeleteTestObject].
+	getImplementation() *assumeCache
+}
+
+// Informer is the subset of [cache.SharedInformer] that NewAssumeCache depends upon.
```

Comment:
Do we have the ShardInformer now?

Reply:
This might belong into my follow-up PR. I'll remove if not needed here.